File: AsyncOperation.m
//
//  AsyncOperation.m
//
//  Created by younggi.lee on 11/07/2018.
//  Copyright © 2018 younggi.lee. All rights reserved.
//

#import "AsyncOperation.h"

typedef NS_ENUM(NSInteger, AsyncOperationState) {
    AsyncOperationStateReady = 0,
    AsyncOperationStateExcuting,
    AsyncOperationStateCanceled,
    AsyncOperationStateFinished
};

@protocol AsyncOperationDelegate <NSObject>

- (void)operationDidFinish:(AsyncOperation *)operation;

@end

@interface AsyncOperation ()

@property (assign, nonatomic) AsyncOperationState state;
@property (weak, nonatomic) id<AsyncOperationDelegate> delegate;

@end


@implementation AsyncOperation

- (BOOL)isRunning {
    return self.state == AsyncOperationStateExcuting;
}

- (BOOL)isCancelled {
    return self.state == AsyncOperationStateCanceled;
}

- (void)start {
    if([self isCancelled]) {
        [self finish];
        return ;
    }
    self.state = AsyncOperationStateExcuting;
    [self mainTask];
}

- (void)cancel {
    self.state = AsyncOperationStateCanceled;
}

- (void)finish {
    self.state = AsyncOperationStateFinished;
    
    if(nil != self.delegate && [self.delegate respondsToSelector:@selector(operationDidFinish:)]){
        [self.delegate operationDidFinish:self];
    }
}

# pragma mark - <AsyncOperationProtocol>

// TODO this method will be excueted when it added although waiting previous operations
- (void)whenQueued {
}

// TODO process and call [finish] method when it done
- (void)mainTask {
    [self finish];
}

@end

@interface AsyncOperationQueue () <AsyncOperationDelegate>

@property (strong) NSMutableArray <AsyncOperation *> *operations;
@property (strong, nonatomic) dispatch_queue_t operationThread;

@end

@implementation AsyncOperationQueue

- (instancetype)init {
    if(self = [super init]) {
        self.operations = [NSMutableArray array];
        self.operationThread = dispatch_queue_create([@"AsyncOperationQueueThread" UTF8String], DISPATCH_QUEUE_SERIAL);
    }
    return self;
}

- (NSUInteger)operationCount {
    return self.operations.count;
}

- (void)addOperation:(AsyncOperation *)operation {
    
    __weak __typeof(self) weakSelf = self;
    dispatch_async(self.operationThread, ^{
        __strong __typeof(weakSelf) strongSelf = weakSelf;
        
        BOOL shouldStartImmediately = (strongSelf.operationCount == 0);
        
        [strongSelf.operations addObject:operation];
        [operation whenQueued];
        
        if(shouldStartImmediately) [strongSelf start];
    });
}

- (void)cancelAllOperations {
    
    AsyncOperation *firstOperation = self.operations.firstObject;
    BOOL shouldFinishFirstObject = firstOperation.state == AsyncOperationStateExcuting;
    
    for(AsyncOperation *operation in self.operations) { [operation cancel]; }
    
    if(shouldFinishFirstObject) [firstOperation finish];
}

- (void)start {
    
    __weak __typeof(self) weakSelf = self;
    dispatch_async(self.operationThread, ^{
        __strong __typeof(weakSelf) strongSelf = weakSelf;
        
        AsyncOperation *operation = strongSelf.operations.firstObject;
        if(nil != operation) {
            operation.delegate = strongSelf;
            [operation start];
        }
    });
}

# pragma mark - <AsyncOperationDelegate>

- (void)operationDidFinish:(AsyncOperation *)operation {
    
    __weak __typeof(self) weakSelf = self;
    dispatch_async(self.operationThread, ^{
        __strong __typeof(weakSelf) strongSelf = weakSelf;
        
        if(strongSelf.operationCount == 0) return ;
        
        if(nil != operation && [strongSelf.operations indexOfObject:operation] != NSNotFound) {
            [strongSelf.operations removeObject:operation];
        }
    });
    [self start];
}

@end

File: AsyncOperation.swift
//
//  AsyncOperation.swift
//  test2
//
//  Created by SKT1110984 on 23/11/2018.
//  Copyright © 2018 SK Telecom. All rights reserved.
//

import Foundation

fileprivate protocol AsyncOperationDelegate: class {
    func didFinish(operation: AsyncOperation);
}

internal enum AsyncOperationState: Int32 {
    case ready
    case executing
    case cancelled
    case finished
}

public class AsyncOperation: NSObject {
    
    fileprivate var state: AsyncOperationState = .ready
    fileprivate weak var delegate: AsyncOperationDelegate?
    
    var running: Bool {
        return state == .executing
    }
    
    var cancelled: Bool {
        return state == .cancelled
    }
    
    public func finish() {
        state = .finished
        delegate?.didFinish(operation: self)
    }
    
    fileprivate func start() {
        if cancelled {
            finish()
            return
        }
        state = .executing
        mainTask()
    }
    
    fileprivate func cancel() {
        state = .cancelled
    }
}

extension AsyncOperation {
    
    // TODO this method will be excueted when it added on queue although waiting previous operations.
    @objc internal func whenQueued() {
    }
    
    // TODO this method will be run when it's turn in queue, and should called -(void)finish method when it done.
    @objc internal func mainTask() {
        self.finish()
    }
}

public class AsyncOperationQueue: NSObject {
    
    fileprivate var operations = NSMutableArray.init()
    fileprivate var operationThread = DispatchQueue(label: "AsyncOperationQueue")
    
    var operationCount: Int {
        return operations.count
    }
    
    public override init() {
    }
    
    public func add(operation: AsyncOperation) {
        operationThread.async {
            let shouldStartImmediately = self.operationCount == 0
            
            self.operations.add(operation)
            operation.whenQueued()
            
            if shouldStartImmediately { self.start() }
        }
    }
    
    public func cancelAll() {
        
        guard let firstOperation = operations.firstObject as? AsyncOperation else { return }
        let shouldFinishFirstObject = firstOperation.state == .executing
        
        for o in operations {
            guard let operation = o as? AsyncOperation else { return }
            operation.cancel()
        }
        if shouldFinishFirstObject { firstOperation.finish() }
    }
    
    private func start() {
        operationThread.async {
            guard let operation = self.operations.firstObject as? AsyncOperation else { return }
            operation.delegate = self
            operation.start()
        }
    }
}

extension AsyncOperationQueue: AsyncOperationDelegate {
    
    fileprivate func didFinish(operation: AsyncOperation) {
        operationThread.async {
            if self.operationCount == 0, self.operations.index(of: operation) == NSNotFound { return }
            self.operations.remove(operation)
            self.start()
        }
    }
}


